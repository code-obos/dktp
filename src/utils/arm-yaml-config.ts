import * as YAML from 'yaml';
import { ARMYamlTemplate } from './arm-yaml-config.types';

export class ArmYamlConfig {
    static async from(file: string): Promise<ArmYamlConfig> {
        const content = await Bun.file(file).text();
        return new ArmYamlConfig(content);
    }
    private content: string;
    private config: ARMYamlTemplate | null;

    constructor(yamlContent: string) {
        this.content = yamlContent;
        this.config = null;
    }

    getConfig(): ARMYamlTemplate {
        if (this.config == null) {
            this.config = YAML.parse(this.content) as ARMYamlTemplate;
        }
        return this.config;
    }

    interpolateWithEnvValues(env: NodeJS.ProcessEnv): ArmYamlConfig {
        this.config = null;
        this.content = this.content.replaceAll(/\$\{\s*([^-}]+(?:-[^}]+)?)\s*\}/g, (_, match) => {
            let key: string = match;
            let defaultValue: string | undefined = match;
            if (match.includes('-')) {
                [key, defaultValue] = match.split('-');
            }

            return env[key] ?? defaultValue ?? key;
        });

        return this;
    }
}
