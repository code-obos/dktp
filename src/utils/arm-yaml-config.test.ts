import { describe, expect, it } from 'vitest';
import { ArmYamlConfig } from './arm-yaml-config';

const yaml = `
properties:
  configuration:
    secrets:
      - name: secret-name
        keyVaultUrl: https://\${KEYVAULT_NAME}.url-to-keyvault.com
        identity: System
  template:
    containers:
      - name: first-container
        resources:
          cpu: \${cpu-0.5}
          memory: \${memory-1.0Gi}
        env:
          - name: ENV
            value: testing
          - name: PASSWORD
            secretRef: secret-name
`;

describe('arm-yaml-test', () => {
    it('should return parsed config', () => {
        const config = new ArmYamlConfig(yaml);
        expectMatchSnapshot(config.getConfig());
    });

    it('should inject values into string, and use default values if needed', () => {
        const config = new ArmYamlConfig(yaml)
            .interpolateWithEnvValues({
                KEYVAULT_NAME: 'mysecrets',
            })
            .getConfig();

        const cpu = config.properties.template.containers.at(0)?.resources?.cpu;
        const memory = config.properties.template.containers.at(0)?.resources?.memory;
        const keyvaultUrl = config.properties.configuration.secrets.at(0)?.keyVaultUrl;
        expect(cpu).toBe(0.5);
        expect(memory).toBe('1.0Gi');
        expect(keyvaultUrl).toBe('https://mysecrets.url-to-keyvault.com');
        expectMatchSnapshot(config);
    });

    it('should inject values into string, and prefer provided values', () => {
        const config = new ArmYamlConfig(yaml)
            .interpolateWithEnvValues({
                cpu: '1.0',
                memory: '1.5Gi',
                KEYVAULT_NAME: 'mysecrets',
            })
            .getConfig();

        const cpu = config.properties.template.containers.at(0)?.resources?.cpu;
        const memory = config.properties.template.containers.at(0)?.resources?.memory;
        const keyvaultUrl = config.properties.configuration.secrets.at(0)?.keyVaultUrl;
        expect(cpu).toBe(1.0);
        expect(memory).toBe('1.5Gi');
        expect(keyvaultUrl).toBe('https://mysecrets.url-to-keyvault.com');
        expectMatchSnapshot(config);
    });

    it('should variable name as default if not specified with a default', () => {
        const config = new ArmYamlConfig(yaml).interpolateWithEnvValues({}).getConfig();

        const keyvaultUrl = config.properties.configuration.secrets.at(0)?.keyVaultUrl;
        expect(keyvaultUrl).toBe('https://${KEYVAULT_NAME}.url-to-keyvault.com');
        expectMatchSnapshot(config);
    });

    it('should work with process.env', () => {
        const yaml = `value: \${PATH}`;
        const config = new ArmYamlConfig(yaml).interpolateWithEnvValues(process.env).getConfig() as any;

        expect(config.value.length).toBeGreaterThan(20);
    });
});

/**
 * Due to bug in bun
 * https://github.com/oven-sh/bun/issues/5272
 */
function expectMatchSnapshot(value: any) {
    const serialized = JSON.stringify(value, null, 2);
    const sanitized = serialized.replaceAll('$', '~');
    expect(sanitized).toMatchSnapshot();
}
