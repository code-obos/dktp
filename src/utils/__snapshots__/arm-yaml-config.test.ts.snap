// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`arm-yaml-test should return parsed config 1`] = `
"{
  "properties": {
    "configuration": {
      "secrets": [
        {
          "name": "secret-name",
          "keyVaultUrl": "https://~{KEYVAULT_NAME}.url-to-keyvault.com",
          "identity": "System"
        }
      ]
    },
    "template": {
      "containers": [
        {
          "name": "first-container",
          "resources": {
            "cpu": "~{cpu-0.5}",
            "memory": "~{memory-1.0Gi}"
          },
          "env": [
            {
              "name": "ENV",
              "value": "testing"
            },
            {
              "name": "PASSWORD",
              "secretRef": "secret-name"
            }
          ]
        }
      ]
    }
  }
}"
`;

exports[`arm-yaml-test should inject values into string, and use default values if needed 1`] = `
"{
  "properties": {
    "configuration": {
      "secrets": [
        {
          "name": "secret-name",
          "keyVaultUrl": "https://mysecrets.url-to-keyvault.com",
          "identity": "System"
        }
      ]
    },
    "template": {
      "containers": [
        {
          "name": "first-container",
          "resources": {
            "cpu": 0.5,
            "memory": "1.0Gi"
          },
          "env": [
            {
              "name": "ENV",
              "value": "testing"
            },
            {
              "name": "PASSWORD",
              "secretRef": "secret-name"
            }
          ]
        }
      ]
    }
  }
}"
`;

exports[`arm-yaml-test should inject values into string, and prefer provided values 1`] = `
"{
  "properties": {
    "configuration": {
      "secrets": [
        {
          "name": "secret-name",
          "keyVaultUrl": "https://mysecrets.url-to-keyvault.com",
          "identity": "System"
        }
      ]
    },
    "template": {
      "containers": [
        {
          "name": "first-container",
          "resources": {
            "cpu": 1,
            "memory": "1.5Gi"
          },
          "env": [
            {
              "name": "ENV",
              "value": "testing"
            },
            {
              "name": "PASSWORD",
              "secretRef": "secret-name"
            }
          ]
        }
      ]
    }
  }
}"
`;

exports[`arm-yaml-test should variable name as default if not specified with a default 1`] = `
"{
  "properties": {
    "configuration": {
      "secrets": [
        {
          "name": "secret-name",
          "keyVaultUrl": "https://KEYVAULT_NAME.url-to-keyvault.com",
          "identity": "System"
        }
      ]
    },
    "template": {
      "containers": [
        {
          "name": "first-container",
          "resources": {
            "cpu": 0.5,
            "memory": "1.0Gi"
          },
          "env": [
            {
              "name": "ENV",
              "value": "testing"
            },
            {
              "name": "PASSWORD",
              "secretRef": "secret-name"
            }
          ]
        }
      ]
    }
  }
}"
`;
